vcycle.conf.5
.TH vcycle.conf 5 "Apr 2018" "vcycle.conf" "vcycle Manual"
.SH NAME
.B vcycle.conf
\- vcycle configuration file
.SH DESCRIPTION
.B vcycled
is a daemon  which implements VM lifecycle management on IaaS cloud systems
in a way inspired by Vac. vcycled reads its configuration from
.B /etc/vcycle.conf
and .conf files in
.B /etc/vcycle.d 
and
.B /var/lib/vcycle/shared/vcycle.d

The configuration files use the Python ConfigParser syntax, which is similar
to MS Windows INI files. The files are divided into sections, with each section
name in square brackets. Each section contains
a series of option=value pairs. Sections with the same name are merged
and if options are duplicated, later values overwrite values given
earlier.

For ease of management, any configuration file ending in .conf in the
directories /var/lib/vcycle/shared/vcycle.d and then /etc/vcycle.d will be read,
in alphanumeric order by name, and then /etc/vcycle.conf is read if present.
 
.SH [SPACE ...] SECTIONS

One [space ...] section must exist for each project, tenancy, or account in which
VMs will be created, with the Vcycle space name for the space given in the section
name, such as [space project1.example.com]. The Vcycle admin has a free choice
of what names to give each space, but the name must only consist of lowercase
letters, numbers, periods, and hyphens so it can be used as the DNS name of
a virtual CE elsewhere in the system.

.B api
is required and specifies the lowercase name of the API to use when contacting
the cloud service associted with this space. API names can consist of letters,
numbers, and underscores.

.B api_version
is optional and requests a particular version of the API understood by its
plugin. API versions can consist of lowercase letters, numbers, hyphens,
periods, and underscores.

.B processors_limit
prevents new machines from being created if the number of processors for
this space in any state exceeds the given limit.

.B flavor_names
gives a space-separated list of one or more names of flavors defined
by the cloud system, which may be used by machinetypes within this space.

.B volume_gb_per_processor
causes Vcycle to create disk volumes of the given size and attach them
to VMs. The volumes are automatically deleted with the VMs. Default 0,
which disables this features.

.B gocdb_sitename
gives the GOCDB site name to use when writing APEL
accounting record files to /var/lib/vcycle/apel-outgoing and
/var/lib/vcycle/apel-archive. Please use your official site name to avoid
the risk of misnamed records getting into the central APEL database.
If gocdb_sitename is not given, then records are only written to
apel-archive and the Vcycle space name is used as a placeholder in the
files.

.B vacmon_hostport
If set, this option gives a space-separated list of HOST:PORT to send
VacQuery UDP messages to. This can be used to monitor the ongoing status
of the Vcycle factory and its VMs via site or central VacMon services.
The central GridPP VacMon service is vacmon.gridpp.ac.uk:8884

.B https_host
gives the FQDN used to contact the Vcycle HTTPS server from
within the VMs. This may be used to direct requests to an HTTPS server
which obtains the files via a shared filesystem. The default value is
the host on which Vcycle runs.

.B https_port
gives the port number used to contact the Vcycle HTTPS server from
within the VMs. This only needs to be changed if there is an intervening
firewall. Default 443.

.B shutdown_time
gives a time to set shutdowntime_job files to, in Unix time. By default these
are just set to be time + max_wallclock_seconds. If this time is in the past,
vcycle stops creating machines for the space.

.B cleanup_hours
gives how many hours to keep per-machine directories in 
/var/lib/vcycle/shared/spaces/SPACE/deleted . The modification time
of the machine's directory is used in the calculation. Default 72.

.SH OPENSTACK SPACE SECTIONS

OpenStack spaces are enabled with
.B api = openstack
and the following options are then available.

.B glance_api
is the glance api version.

.B url
is the URL of the identity (KeyStone) endpoint for this OpenStack service.

.B project_name
is the project or tenancy name.

.B domain_name
is the domain name containing the project or 'default' by default.

.B region
,
.B zones
and
.B security_groups
optionally give the OpenStack region, zone(s) and security_group(s)
to use when creating VMs.

.B network_uuid
gives the UUID of a network to use. By default no network is requested
and OpenStack uses the default network for the project.

.B username
is a user with access to the project.

.B password_base64
is the base64 encoded password for the user. Vcycle will decode the
value to obtain the true password. This encoding is to avoid revealing
memorable passwords when casually viewing or editing configuration files.
You can encode the real password with something like:
.br
.ce
echo -n 'PASSword' | base64

.B cred_id
and
.B cred_secret
give an application credential ID and secret as an alternative to 
username and password.

When creating VMs in OpenStack spaces, Vcycle will create "machinefeatures",
"jobfeatures", and "joboutputs" metadata keys with the URLs of the
corresponding directories for the VM on the Vcycle machine's HTTP(S)
server.

.SH EC2 SPACE SECTIONS

As well as OpenStack, Vcycle natively supports the EC2 API. However, you
should use the OpenStack API if available. Spaces accessed with the EC2 API
are enabled with
.B api = ec2
and the following options are then available.

.B url
is the URL of the endpoint of this EC2 service.

.B access_key
and
.B secret_key
are the credentials needed to access the service.

.B version
explicitly gives the API version to be passed to EC2. Default "2010-08-31".

.B region
explicitly gives the region to be passed to EC2. Default "us-east-1".

.B service
explicitly gives the service name to be passed to EC2. Default "openstack".

When creating VMs in EC2 spaces, Vcycle will create "machinefeatures",
"jobfeatures", and "joboutputs" metadata tags with the URLs of the
corresponding directories for the VM on the Vcycle machine's HTTP(S)
server.

.SH MACHINE / JOB FEATURES HTTP(S) SERVER

Vcycle will create the machinefeatures and jobfeatures files according
to the WLCG MJF protocol described in HSF-TN-2016-02 in machinefeatures and
jobfeatures subdirectories
of the VM's directory in /var/lib/vcycle/shared/spaces/SPACE/current . 
Using the vcycle.httpd.inc
Apache configuration file supplied in /usr/share/doc/vcycle-VERSION, these
files are publish via HTTP and HTTPS.

In addition, if https_x509dn is set for the VM's machinetype, then the script vcycle-cgi
will allow the VM to write heartbeat, logging, and shutdown_message files
to the VM's $JOBOUTPUTS directory.

This server also publishes the heartbeat machines lists of the machines with
a current heartbeat for each machinetype in each space described in the
.B USER_DATA SUBSTITUTIONS
section.

.SH [MACHINETYPE ... ...] SECTIONS

One [machinetype ... ...] section must exist for each machinetype in each space, with
the space name and name of the machinetype given in the section name, such as
[machinetype project1.example.com example].
A machinetype name must only consist of lowercase letters, numbers, and hyphens.
Each of these sections contain option=value pairs that are specific to
that machinetype. The same machinetype name can appear in different spaces and will
be managed separately.

.B flavor_names
gives a space-separated list of one or more names of flavors defined
by the cloud system which represent particular combinations of CPU, memory, and
disk geometry. This option overrides any flavors list given in the space
section. Subject to any constraints imposed by
.B min_processors
and
.B max_processors
the first allowed flavor in the list is used.

.B min_processors 
and
.B max_processors
give the minimum and maximum number of logical processors which can be allocated
to LMs of this type when they are created. Vcycle will attempt to select a 
flavor from the flavor_name list matching these constraints. If the API plugin
cannot determine the number of processors allocated to a VM, then Vcycle will use 
.B min_processors
in its calculations and limits, and when setting the values of 
$MACHINEFEATURES/total_cpu and $JOBFEATURES/allocated_cpu supplied to the VM.
The default minimum is 1 and the default is no maximum.

.B hs06_per_processor
gives the HEPSPEC06 power of each processor in the virtual machines created
for this flavor in this machinetype. If set, this is used to calculate the value
$MACHINEFEATURES/hs06 and $JOBFEATURES/hs06_job
supplied to the VM. It is also used when calculating target shares and
when writing APEL accounting records, and for both of these a default of 1.0
is used if not set explicitly.

.B mb_per_processor
gives the number MB of each virtual machine created for this flavor
in this machinetype. This is used as the value of $JOBFEATURES/max_rss_bytes
supplied to the VM. If the api plugin can
positively determine the number from metadata about the flavor, it will be
used in preference to the value given here. Default 2048 per processor.

.B target_share
gives the desired share of the capacity available in this space for this
machinetype. The shares do not need to add up to 1.0, and if a share is not given
for a machinetype, then it is set to 0. Vcycle consults these shares
when deciding which machinetype to start as VM capacity becomes available.
Shares are weighted by the hs06 value of the machinetype.

.B backoff_seconds
is the delay after a VM of this machinetype aborts. If a VM aborts, then no new
VMs of this type will be created for this amount of time. This can be used
to prevent the unnecessary creation of many VMs when no work is available,
and avoid overloading the matcher or task queue of the VO.

.B fizzle_seconds
is used in three places within the backoff procedure and in two
other parts of Vcycle:
.br
(1) First, if a VM finishes
without producing a shutdown message code and has lasted less than
fizzle_seconds, then it is treated as aborted.
.br
(2) Secondly, after the
backoff_seconds time has expired for a VM abort, once at least one VM has
been started in this Vcycle space, then no more new VMs can be started for
another fizzle_seconds.
.br
(3) Additionally, when writing the accounting log files, any VMs which
run for less than fizzle_seconds are excluded.
.br
(5) Finally, the heartbeat file
checking is only carried out once an initial period of fizzle_seconds
has passed.

.B accounting_fqan
is used to specify a FQAN to include when writing APEL accounting
records, to associate usage with particular experiments.

.B processors_limit
prevents new machines from being created if the number of processors for
this machinetype in any state exceeds the given limit.

.B max_wallclock_seconds
gives the maximum lifetime of a VM. Vcycle will create
$MACHINEFEATURES/shutdowntime inside the VM using this value to
communicate it to the VM. Vcycle will destroy the VM if it is still
running after this amount of time. Default 86400.

.B heartbeat_file
allows the machinetype to nominate a file which will be created in
the $JOBOUTPUTS directory before fizzle_seconds has passed. If this
file is not created by then and maintained for the lifetime of the VM,
the VM will be destroyed.

.B heartbeat_seconds
gives the frequency at which the heartbeat_file must be updated after
fizzle_seconds has passed. If the file is not updated for
heartbeat_seconds then the VM will be destroyed. If heartbeat_seconds
is 0, then only the existence of the file will be checked. Default 0.

.B cvmfs_proxy_machinetype
gives the name of another machinetype consisting of HTTP caching proxies
suitable for cvmfs running inside the VMs. The list of proxies is made
available to VMs via the ##user_data_option_cvmfs_proxy## substitution.
If :PORTNUMBER is appended to the name of the machinetype, then that 
port number will be used in the cvmfs proxy lists Vcycle generates.
See 
.B USER_DATA SUBSTITUTIONS
for more details.

.B https_x509dn
is an optional X.509 DN which will be used by the vcycle-cgi script to
control writing to VMs' $JOBOUTPUTS directories on the local HTTPS
server.

.B legacy_proxy
can be set to True to generate Globus legacy proxies rather than RFC 3820
proxies. Default False.

.B user_data_proxy
set to true causes the files x509cert.pem and x509key.pem in the
machinetype's subdirectory of /var/lib/vcycle/spaces/SPACE/machinetypes to
be used to make a limited X.509 proxy. The two files can be
identical if desired, and the X.509 certificate and RSA private key
will be extracted from the files as appropriate. (Note that this location
is one level above the files subdirectory in which the following options
look by default.)

For the remaining options, if the file name begins with '/', then it
will be used as an absolute path; otherwise the path will be interpreted
relative to the machinetype's subdirectory of /var/lib/vcycle/spaces/SPACE/machinetypes/MACHINETYPE/files
where SPACE is the parent space name and MACHINETYPE is the name of
this machinetype.

.B root_image
identifies the image file from which the VM will boot. If the cloud
service already has the desired image, then it can be referenced by
prefixing the service's native image ID with "image:".
.br
For the OpenStack API, root_image can be
the path to the image file itself on the local filesystem. Alternatively,
it can also be a remote HTTP or HTTPS URL which Vcycle
will cache in /var/lib/vcycle/imagecache. The remote server must supply a
Last-Modified timestamp and Vcycle will re-request the image each time a
VM starts using an If-Modified-Since request to minimise network load.
Alternatively, the images may be files in the local filesystem. If
root_image ends in .iso , then the image will be declared as ISO format
(a CD-ROM image), otherwise as a raw HDD image.

.B cernvm_signing_dn
is used to specify a regular expression to match the DN of an X.509
certificate used to verify the authenticity of the root image. Vcycle
attempts to obtain the certificate and signature from a CernVM Signature
Block at the end of the image file, verifies the
certificate using the CA files in /etc/grid-security/certificates, and
compares the certificate DN to cernvm_signing_dn. If this option is
given, all these verification steps must be satisified for the image
to be used. As of 2016, CernVM images are signed with a DN matching
the regular expression /CN=cvm-sign01\\.cern\\.ch$

.B root_public_key
is the file name of a public key which Vcycle will set up on the cloud
system and supply to the VMs to allow root ssh access. If this option is
not given and the file /root/.ssh/id_rsa.pub exists, it will be used
instead, giving ssh access from the factory machine.

.B user_data
is the path of a contextualization file provided by the VO and perhaps
modified by the site. If the path is a remote HTTP or HTTPS URL, Vcycle
will fetch it over the network each time a VM is started. However the
file is obtained, Vcycle will apply a series of default and locally defined
##user_data___## substitutions to it. See USER_DATA SUBSTITUTIONS below
for a list of the default substitutions.

.B user_data_option_XXX
and
.B user_data_file_XXX
are locally defined substitutions which will be applied to the user_data
file before the VM is started. user_data_option_XXX takes the string to
be substituted. user_data_file_XXX takes the relative or absolute path to
a file whose contents will be substituted for the pattern in the
user_data file.

.SH USER_DATA SUBSTITUTIONS

Before the user_data file is used in starting a VM, several pattern based
substitutions are performed by Vcycle. These patterns are in the form
##user_data___##. String values given to the option user_data_option_XXX
replace patterns of the form ##user_data_option_XXX##. The contents of
the files given to user_data_file_XXX options also replace patterns of the
form ##user_data_option_XXX##. In both cases XXX are arbitrary strings
consisting of letters, numbers, and underscores.

The pattern ##user_data_x509_proxy## is replaced by the X.509 proxy 
created if the user_data_proxy_cert and user_data_proxy_key options
are given.

.B cvmfs_proxy_machinetype
may be given in a machinetype definition 
with the name of another machinetype consisting of HTTP 
caching proxies suitable for cvmfs. The list of proxies is made
available to VMs via the normal ##user_data_option_cvmfs_proxy## 
substitution.
Only machines with a valid heartbeat are included, and they are 
identified by IP address in the format http://xxx.xxx.xxx.xxx:ppp and 
separated by pipe ('|')
characters so they are used in round-robin mode by cvmfs. For
security reasons, these proxies are accessed on port 280 by default 
not 3128. This port number may be changed by appending :PORTNUMBER 
to the given machinetype name. 
If user_data_option_cvmfs_proxy is also given, then a semicolon
separator and the option's value are appended to
the list of proxies obtained from the machinetype's list. This allows
a static choice of backup proxies to be given, which will benefit from
cvmfs's failover feature if none of the machinetype proxies are usable.

In addition, the following substitutions are performed automatically by
Vcycle using data it holds internally:

.br
.B ##user_data_space##
is the Vcycle space name.
.br
.B ##user_data_url##
is the HTTP(S) URL from which the user_data template was obtained. Only given if
the template was retrieved by HTTP(S) rather from a local path.
.br
.B ##user_data_machinefeatures_url##
and
.B ##user_data_jobfeatures_url##
and
.B ##user_data_joboutputs_url##
are the values of $MACHINEFEATURES, $JOBFEATURES, and $JOBOUTPUTS to set
within the VM.
.br
.B ##user_data_heartbeat_machines_url##
is the HTTP(S) URL from which a list of the machines in this space with
the same machinetype as this machine can be found. Only machines with a
current heartbeat are included. The list consists of the heartbeat time
in Unix seconds, the machine name, and the local IP address of the machine 
within the remote cloud.
.br
.B ##user_data_machinetype##
is the name of the machinetype of this VM.
.br
.B ##user_data_machine_hostname##
is the hostname given to the VM by Vcycle.
.br
.B ##user_data_manager_version##
has the form "Vcycle v.v.v" where v.v.v is the Vcycle version.
.br
.B ##user_data_manager_hostname##
is the hostname of the machine on which the Vcycle daemon is running.

.SH AUTHOR
Andrew McNab <Andrew.McNab@cern.ch>

vcycled is part of Vcycle: https://www.gridpp.ac.uk/vcycle/
.SH "SEE ALSO"
.BR vcycled(8),
.BR vcycle-cgi(8)


CHANGES
==================== Changes in Vcycle version 03.00.00 =====================
- Remove GOCDB updating mechanism and gocdb_cert/key_file option
- Add min/max_wallclock_seconds, min/max_processors, bytes_per_processor
  to machinetypes VacQuery messages to VacMon
- Fix $MACHINEFEATURES/shutdowntime key name
- Create vcycled logs in /var/log/vcycle to allow symlink redirection
- Move directories for removed machines to deleted machines directory
- Do not try to process machines if fail to connect to space's service
- Remove joboutputs copying and joboutputs configuration options
- Remove remote joboutputs handling
- Seperate hiearchies in /var/lib/vcycle/shared/SPACE for each space
- x509dn becomes https_x509dn, now recorded in each machines directory
- Add https_host to allow a remote HTTPS server using a shared filesystem
- vcycle-cgi now no longer depends on Vcycle or Vacutils modules
- Add mechanism to take machines abandoned by other Vcycle instances
- Read config files in /var/lib/vcycle/shared/vcycle.d too
- Add ##user_data_site##
- Support application credential authentication in OpenStack
==================== Changes in Vcycle version 01.00.01 =====================
- Fix cleanup of machine files from machines in spaces removed from config
==================== Changes in Vcycle version 01.00.00 =====================
- Create 1024bit GSI proxies
- Update MJF for HSF-TN-2016-02
- Remove GLUE2.0/JSON status publishing
- Include admin-guide.html and update example.vcycle.conf
- Update vacutils with ##user_data_url## set to remote user_data URL
- Vacutils supports Vacuum Pipes.
- user_data_proxy_cert/_key replaced by user_data_proxy True/False with
  fixed x509cert.pem/x509key.pem file names.
- user_data and root image URLs in user_data templates
- Send VacMon factory_status and machinetype_status messages every cycle,
  and machine_status when each machine finishes
- Identity V3 and Glance V2 support in OpenStack plugin, 
  tenancy_name -> project_name, and add domain_name option.
- hs06 -> hs06_per_processor; cpu_per_machine -> min_processors
- running_hs06 isn't published to VacMon if no hs06_per_processor
- Add region and zones options to OpenStack plugin
- Create /var/lib/vcycle/SPACENAME/machinetypelists/MACHINETYPENAME lists
  of machines for each machinetype with current heartbeats
- Add cvmfs_proxy_machinetype option which constructs the proxies list 
  user_data_option_cvmfs_proxy from the IP addresses of machines of the
  given machinetype with a current heartbeat (ie a Squid machinetype)
- processors_limit replaces max_machines or max_processors to limit 
  processors per space or per machinetype
- Add gocdb_cert_file/gocdb_key_file options to publish space information
  to GOCDB
==================== Changes in Vcycle version 0.8.0 =====================
- Use spaceDir not spaceName in cleanupJoboutputs()
- machinetypes directories are now under /var/lib/vcycle/spaces/
- Use common Vac Project vcycle.httpd.conf and specific vcycle.httpd.inc
- GLUE2.0/JSON publishing of status
- cernvm_signing_dn option added, to check signatures of CernVM boot
  images
- Force use of TLSv1 (pycurl.SSLVERSION_TLSv1)
- Base64 encoding and password_base64 in OpenStack plugin
- Publish status in GLUE 2.1 Cloud* as well as GLUE 2.0
- EC2 API plugin
==================== Changes in Vcycle version 0.7.0 =====================
- Use machineoutputs -> joboutputs internally and in options
==================== Changes in Vcycle version 0.6.0 =====================
- Add usercert/userkey to httpRequest if need to handle X.509/Proxy/VOMS
- Add usercert/userkey to openstack_api.py as username alternative
- Import azure_api.py and dbce_api.py
- Add vcycle-azure and vcycle-dbce RPM subpackages
- Update revised OCCI plugin from Luis Villazon Esteban
- Rename vmtype -> machinetype in code, configuration, and documentation
- Use vacutils.py with MACHINEFEATURES, JOBFEATURES, and JOBOUTPUTS
  substitutions in user_data
- Add joboutputs URL to metadata in OpenStack plugin
- APEL job records now have SubmitHost: SPACE/vcycle-HOSTNAME where
  HOSTNAME is the host name of this Vcycle instance
- Add cpu_per_machine for platforms where this can't be determined
- Remote $JOBOUTPUTS support
==================== Changes in Vcycle version 0.5.0 =====================
- Create occi_api.py
- Each cycle is run in a newly forked subprocess for stability
- Record APEL accounting messages when VMs finish
- Use hs06 weighting when calculating target shares
- Proper man pages
- OpenStack API can now upload ISO format images (ending in .iso)
- Better guessing of IP addresses of VMs
- https_port option allows setting of MJF HTTPS server port
- Explicitly set the architecture property in uploaded images to x86_64
==================== Changes in Vcycle version 0.4.0 =====================
- VCYCLE.py becomes /usr/lib64/python2.6/site-packages/vcycle
- Put docs in /usr/share/doc/vcycle-VERSION
- Put vcycled, vcycle, vcycle-cgi in /usr/sbin
- Add user_data_proxy_cert/user_data_proxy_key/legacy_proxy options in
  vmtypes to put a proxy in user_data rather than a hostcert/hostkey
- creationsPerCycle calculated as 10% of max_machines in the space
- Recast using BaseSpace, OpenstackSpace, and Vmtype objects
- Use OpenStack REST API directly and drop nova dependency
- Vcycle manages key pairs and boot images automatically
- OpenstackSpace subclass moved to openstack_api.py API plugin module
==================== Changes in Vcycle version 0.3.0 =====================
- Separate [space ...] and [vmtype ...] sections in vcycle.conf
- Handle multiple vmtypes per space
- Per-space user_data directories, containing per-vmtype user_data files
- Use "ROT-1" encoding for passwords in config to avoid accidental
  disclosure if browsing around. Obviously this is trivial to get around
  if one wants to.
- Allow connections to unauthenticated HTTPS servers for now
- Proper cleanup of old machine directories
- Target shares for vmtypes within spaces

VCYCLE.py
#!/usr/bin/python
#
#  VCYCLE.py - vcycle library
#
#  Andrew McNab, University of Manchester.
#  Copyright (c) 2013-4. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or
#  without modification, are permitted provided that the following
#  conditions are met:
#
#    o Redistributions of source code must retain the above
#      copyright notice, this list of conditions and the following
#      disclaimer. 
#    o Redistributions in binary form must reproduce the above
#      copyright notice, this list of conditions and the following
#      disclaimer in the documentation and/or other materials
#      provided with the distribution. 
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
#  CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
#  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
#  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
#  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
#  BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
#  TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
#  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
#  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
#  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
#  POSSIBILITY OF SUCH DAMAGE.
#

import os
import sys
import time
import json
import tempfile
import logging
import ConfigParser

tenancies    = None
lastFizzles = {}
loggers = {}

def readConf(requirePassword=True):

  global tenancies, lastFizzles
  
  tenancies = {}

  tenancyStrOptions = [ 'tenancy_name', 'url', 'username', 'proxy' , 'type', 'auth' ]

  tenancyIntOptions = [ 'max_machines' ]

  vmtypeStrOptions = [ 'ce_name', 'image_name', 'flavor_name', 'root_key_name', 'x509dn', 'network', 'public_key' ]

  vmtypeIntOptions = [ 'max_machines', 'backoff_seconds', 'fizzle_seconds', 'max_wallclock_seconds' ]

  parser = ConfigParser.RawConfigParser()
  
  # Look for configuration files in /etc/vcycle.d
  try:
    confFiles = os.listdir('/etc/vcycle.d')
  except:
    pass 
  else:
    for oneFile in sorted(confFiles):
      if oneFile[-5:] == '.conf':      
        try:
          parser.read('/etc/vcycle.d/' + oneFile)
        except Exception as e:
          logLine('Failed to parse /etc/vcycle.d/' + oneFile + ' (' + str(e) + ')')

  # Standalone configuration file, read last in case of manual overrides
  parser.read('/etc/vcycle.conf')

  # First look for tenancy sections

  for tenancySectionName in parser.sections():
    split1 = tenancySectionName.lower().split(None,1)

    if split1[0] == 'vmtype':    
      continue
      
    elif split1[0] != 'tenancy':
      return 'Section type ' + split1[0] + ' not recognised'
      
    else:
      tenancyName = split1[1]
# NEED TO CHECK THIS IS JUST a-z,0-9,-,_,.
      tenancy = {}
      
      # Get the options from this section for this tenancy
      if not parser.has_option(tenancySectionName, 'tenancy_name') :
         return 'Option tenancy_name required in [' + tenancySectionName + ']'
      
      if not parser.has_option(tenancySectionName, 'url') :
         return 'Option url required in [' + tenancySectionName + ']'
      
      if not parser.has_option(tenancySectionName, 'proxy') and not parser.has_option(tenancySectionName, 'username'):
         return 'Option proxy or username is required in [' + tenancySectionName + ']'
      
      tenancy['tenancy_name'] = parser.get(tenancySectionName,'tenancy_name') 
      tenancy['url'] = parser.get(tenancySectionName,'url')
      tenancy['type'] = parser.get(tenancySectionName,'type')
      
      if not parser.has_option(tenancySectionName, 'auth'):
         tenancy['auth'] = 'x509'
      else:
         tenancy['auth'] = parser.get(tenancySectionName,'auth')
      
      if parser.has_option(tenancySectionName,'proxy'):
         tenancy['proxy'] = parser.get(tenancySectionName,'proxy') 
         requirePassword = False
      else:
         tenancy['username'] = parser.get(tenancySectionName,'username') 
      
      
      for opt in tenancyIntOptions:
        try:
          tenancy[opt] = int(parser.get(tenancySectionName, opt))
        except:
          return 'Option ' + opt + ' required in [' + tenancySectionName + ']'

      try:
        tenancy['password'] = parser.get(tenancySectionName, 'password')
      except:
        if requirePassword:
          return 'Option password is required in [' + tenancySectionName + ']'
        else:
          tenancy['password'] = ''

      try:
         tenancy['delete_old_files'] = bool(parser.get(tenancySectionName, 'delete_old_files'))
      except:
         tenancy['delete_old_files'] = True
         
      try:
         tenancy['delete_no_tenancy'] = bool(parser.get(tenancySectionName, 'delete_no_tenancy'))
      except:
         tenancy['delete_no_tenancy'] = False  

      # Get the options for each vmtype section associated with this tenancy

      vmtypes = {}

      for vmtypeSectionName in parser.sections():
        split2 = vmtypeSectionName.lower().split(None,2)

        if split2[0] == 'vmtype':

          if split2[1] == tenancyName:
            vmtypeName = split2[2]
            # NEED TO CHECK THIS IS JUST a-z,0-9,-,_,.
            vmtype = {}

            for opt in vmtypeStrOptions:              
              if parser.has_option(vmtypeSectionName, opt) :
                vmtype[opt] = parser.get(vmtypeSectionName, opt)
              else:
                if opt is 'network' or 'public_key':
                   continue
                return 'Option ' + opt + ' required in [' + vmtypeSectionName + ']'

            for opt in vmtypeIntOptions:
              try:
                vmtype[opt] = int(parser.get(vmtypeSectionName, opt))
              except:
                return 'Option ' + opt + ' required in [' + vmtypeSectionName + ']'

            try:
              vmtype['heartbeat_file'] = parser.get(vmtypeSectionName, 'heartbeat_file')
            except:
              pass

            try:
              vmtype['heartbeat_seconds'] = int(parser.get(vmtypeSectionName, 'heartbeat_seconds'))
            except:
              pass
           
            try:
              vmtype['username'] = parser.get(vmtypeSectionName, 'username')
            except:
              pass
           
            try:
              vmtype['password'] = parser.get(vmtypeSectionName, 'password')
            except:
              pass
           
            if tenancyName not in lastFizzles:
              lastFizzles[tenancyName] = {}
              
            if vmtypeName not in lastFizzles[tenancyName]:
              lastFizzles[tenancyName][vmtypeName] = int(time.time()) - vmtype['backoff_seconds']

            vmtypes[vmtypeName] = vmtype

      if len(vmtypes) < 1:
        return 'No vmtypes defined for tenancy ' + tenancyName + ' - each tenancy must have at least one vmtype'

      tenancy['vmtypes']     = vmtypes
      tenancies[tenancyName] = tenancy

  return None

def createFile(targetname, contents, mode=None):
  # Create a text file containing contents in the vcycle tmp directory
  # then move it into place. Rename is an atomic operation in POSIX,
  # including situations where targetname already exists.
   
  try:
    ftup = tempfile.mkstemp(prefix='/var/lib/vcycle/tmp/temp',text=True)
    os.write(ftup[0], contents)
       
    if mode: 
      os.fchmod(ftup[0], mode)

    os.close(ftup[0])
    os.rename(ftup[1], targetname)
    return True
  except:
    return False

def makeJsonFile(targetDirectory):
  # Create a dictionary containing the keys and values from the given directory
  # and then write to .json in that directory
  
  outputDict = {}
  
  try:
    filesList = os.listdir(targetDirectory)
  except Exception as e:
         logLine('Listing directory ' + targetDirectory + ' fails with ' + str(e))
  else:
    for oneFile in filesList:
     if oneFile[0] != '.' and not os.path.isdir(targetDirectory + '/' + oneFile):
       try:
         outputDict[oneFile] = open(targetDirectory + '/' + oneFile).read()
       except Exception as e:
         logLine('Failed reading ' + targetDirectory + '/' + oneFile + ' with ' + str(e))
         pass

  try:
    f = open(targetDirectory + '/.json', 'w')
    json.dump(outputDict, f)
    f.close()   
  except Exception as e:
    logLine('Writing JSON fails with ' + str(e))

def logLine(tenancy, text):
  global loggers
  if not tenancy in loggers:
     logger = logging.getLogger(tenancy)
     logger.propagate = False
     logger.setLevel(logging.DEBUG)
     # create file handler which logs even debug messages
     fh = logging.FileHandler("/var/log/vcycle-%s.log" % tenancy)
     fh.setLevel(logging.DEBUG)
     formatter = logging.Formatter('%(asctime)s - %(message)s')
     fh.setFormatter(formatter)
     logger.addHandler(fh)
     loggers[tenancy] = logger

  loggers[tenancy].debug(text)
  #sys.stderr.write(time.strftime('%b %d %H:%M:%S [') + str(os.getpid()) + ']: ' + text + '\n')
  #sys.stderr.flush()
